### DISCUSSION TITLE
Ability to add new extrahost (new line into /etc/hosts file) to already created container

### DISCUSSION THREAD

**Discussion by igribkov** [testcontainers/testcontainers-go#1906] [Q&A] [answered]
Could somebody help me please?
I have already created container
In my create request `testcontainers.ContainerRequest` i've pass an extrahost, for example `"some_host:192.168.1.11"`
then i've call `testcontainers.GenericContainer`
and by some reasons before i will start this container i need to add one more extrahost
So the question is how to do this?

Same question about adding some extra env variables

**Comment by mdelapenya** [comment]
Hey @igribkov thanks for sending this question. It's possible to set the extrahost of a container leveraging the `HostConfigModifier` function in the ContainerRequest struct. Same for environment variables. I'm adding an example test that works:

```golang
func TestDockerContainerWithExtraHosts(t *testing.T) {
	ctx := context.Background()

	nginxC, err := GenericContainer(ctx, GenericContainerRequest{
		ProviderType: providerType,
		ContainerRequest: ContainerRequest{
			Image:        nginxAlpineImage,
			ExposedPorts: []string{nginxDefaultPort},
			Env:          map[string]string{"FOO": "BAR"},
			WaitingFor:   wait.ForListeningPort(nginxDefaultPort),
			HostConfigModifier: func(hc *container.HostConfig) {
				hc.ExtraHosts = []string{"DEV01:172.16.0.10", "DEV01:172.16.0.11"}
			},
		},
		Started: true,
	})

	require.NoError(t, err)
	terminateContainerOnEnd(t, ctx, nginxC)

	_, reader, err := nginxC.Exec(ctx, []string{"cat", "/etc/hosts"}, tcexec.Multiplexed())
	require.NoError(t, err)

	hostsBytes, err := io.ReadAll(reader)
	require.NoError(t, err)

	hosts := string(hostsBytes)

	assert.Contains(t, string(hosts), "172.16.0.10\tDEV01")
	assert.Contains(t, string(hosts), "172.16.0.11\tDEV01")
}
```

Please pay attention to the Env HostConfigModifier fields in the request struct 👀

**Reply by igribkov** [reply]
@mdelapenya tnx for reply

In your example you show how to add `extrahost` to container on creating request

could you please help to add `extrahosts` to container after it has been created
in your example let's add some code to add `extrahosts` after last `assert.Contains`

**Reply by mdelapenya** [reply, marked as answer]
@igribkov ah ok, I'm sorry, I thought your question was about container creation instead of "after it's been created and its running", right?

You can leverage the `PostStarts` container lifecycle hook, to inject a function that appends the entries you are interested. Please check the logs for the lifecycle hooks [here](https://golang.testcontainers.org/features/creating_container/#lifecycle-hooks), and an example here: https://github.com/testcontainers/testcontainers-go/blob/5562dbcfc3ba37d636870f8083cd2d5b40c9e832/options_test.go#L131-L162.

The above snippet is creating a `.testcontainers` file in the `/tmp` dir leveraging the `WithStartupCommand` functional option. This option will execute the commands **right after the container has started, but before your own test** code, because it's coupled with the container lifecycle.

If you want to add an entry after some assertions, which could purely represent a different test case, you could directly call the `container.Exec` method to modify the file with the same commands you could have used in the above alternative. A valid solution for this could be reading the file from outside the container, modifying the bytes in the tests, and copying them back to the container.

**Reply by igribkov** [reply]
@mdelapenya, tnx a lot for your response. \
`testcontainers.WithStartupCommand()` that's exactly what i need

**Reply by mdelapenya** [reply]
At the end of the day, the WithStartupCommand calls the container.Exec method 😉

**Comment by mdelapenya** [comment]
I'm closing this discussion as resolved, thanks for asking!
